internal/controller/linodemachine_controller_helpers.go

```diff
@@ -553,6 +537,81 @@ func setUserData(ctx context.Context, machineScope *scope.MachineScope, createCo
 	return nil
 }
 
+func resolveBootstrapData(ctx context.Context, machineScope *scope.MachineScope, gzipCompressionEnabled bool, logger logr.Logger) ([]byte, error) {
+	bootstrapdata, err := machineScope.GetBootstrapData(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		size       = len(bootstrapdata)
+		compressed []byte
+		limit      int
+	)
+
+	// Determine limits for delivery method, e.g. Metadata vs. Stackscript.
+	if machineScope.LinodeMachine.Status.CloudinitMetadataSupport {
+		limit = maxBootstrapDataBytesCloudInit
+	} else {
+		limit = maxBootstrapDataBytesStackscript
+	}
+
+	switch {
+	// Best case: Can just deliver user-data directly.
+	case size < limit:
+		return bootstrapdata, nil
+	// Compromise case (Metadata): Use compression.
+	case machineScope.LinodeMachine.Status.CloudinitMetadataSupport && gzipCompressionEnabled:
+		compressed, err = compressUserData(bootstrapdata)
+		// Fallback to Object Storage workaround on compression failure.
+		if err != nil {
+			logger.Info(fmt.Sprintf("Failed to compress bootstrap data: %v", err))
+			logger.Info("Falling back to Object Storage")
+			goto obj
```

Comment:
Can we instead have a helper function called something like `storeMetadataInObj`?

Reply:
Yeah, I can pull out the logic _after_ this entire switch into it's own helper function. 

However, is the aim to remove the `goto` statements here? I think we still need to keep this since the aim of this logic is to bail out early when we know we're forced to used the Cluster Object Store workaround.

Reply:
Yeah I meant doing something like this:
```
	case machineScope.LinodeMachine.Status.CloudinitMetadataSupport && gzipCompressionEnabled:
		compressed, err = compressUserData(bootstrapdata)
		// Fallback to Object Storage workaround on compression failure.
		if err != nil {
			logger.Info(fmt.Sprintf("Failed to compress bootstrap data: %v", err))
			logger.Info("Falling back to Object Storage")
			return storeMetadataInObj(ctx, machineScope, logger, bootstrapdata)
		}
		size = len(compressed)
		if size < limit {
			return compressed, nil
		}
		return storeMetadataInObj(ctx, machineScope, logger, bootstrapdata)
	// Worst case: Object Storage workaround.
	default:
		logger.Info("decoded bootstrap data exceeds size limit", "limit", limit, "size", size)
		return storeMetadataInObj(ctx, machineScope, logger, bootstrapdata)
	}
```

Reply:
I reworked the logic inside the switch statement to:
- Reduce repetition between all the different edge-cases and preserve the overall "layered" visual structure
- Make the (now single) `goto` statement function more like a `break` out of the nested switch logic and the continue executing the rest of the function